### ISSUE TITLE
SFTP repository fails without posix-rename extension

### ISSUE THREAD

**Issue by sjend** [kopia/kopia#1099]
The current implementation of the `PutBlobInPath` method in the SFTP repository relies on calling `s.cli.PosixRename`, which uses the (nonstandard) `posix-rename@openssh.com` SFTP extension.

https://github.com/kopia/kopia/blob/6811f1fae1f70bea77fce9858d20a821c41d8601/repo/blob/sftp/sftp_storage.go#L132-L139

The cloud storage provider I am trying to use does not support this extension for some reason. As a result, both creating a new repository and creating a snapshot when connected to an existing repository failed (specifically, the `PutBlob` operation failed with `SSH_FX_OP_UNSUPPORTED`).

From what I could gather

```go
s.cli.PosixRename(tempFile, fullPath)
```
appears to be functionally equivalent to unconditional remove + rename:

```go
s.cli.Remove(fullPath)
s.cli.Rename(tempFile, fullPath)
```

It should also be possible to check for support of the extension using

```go
s.cli.HasExtension("posix-rename@openssh.com")
```
and only use remove+rename as a fallback if `posix-rename` is not supported.

I am unsure of the performance implications of unconditionally removing a file or possible correctness issues with removing a non-existent file or any other potential issues, but replacing the call to `PosixRename` with the other two calls at least allowed me to create a repository and a snapshot.

If it is not possible to create a fallback, it may be worth pointing out that support for the `posix-rename` extension is required to use the SFTP backend (though admittedly, I don't know if lack of support for this extension is common or I just got unlucky with a peculiar setup from my provider).

**Comment by jkowalski** [comment]
Because Kopia never needs to replace files in a repository during normal operation, that approach should be equivalent. I'll be happy to review a PR.

**Comment by sjend** [comment]
Thank you for your quick response and all the work on Kopia. I can try to get you a PR, depending on what we decide.

> Kopia never needs to replace files in a repository during normal operation

This sounds interesting. If we know that the files don't exist, then we can skip removing the file beforehand and just use `s.cli.Rename(tempFile, fullPath)` directly.
I decided to look into this and it turns out that the initial implementation ([here](https://github.com/jbrodriguez/kopia/blob/30890ce5cac483276566bc3eaed31af47f270ed4/repo/blob/sftp/sftp_storage.go#L102), merged with #55) did just that and was changed later (in #68) due to the tests failing.

I am not entirely sure on how to proceed from here. The best solution to me seems to be to restore the original behaviour (i.e. use `s.cli.Rename(tempFile, fullPath)`) and to change the test ([I think it's this one](https://github.com/kopia/kopia/blob/e0e7f66a90e6c25120fb3dcd4db1df223041147e/internal/blobtesting/verify.go#L87)), if there indeed is no requirement to overwrite blobs (which has the benefit of aligning the asserted functionality with the used functionality). Such a change would have a wider impact and goes a bit beyond what I feel comfortable doing with Go, so I would appreciate your guidance on that.
Alternatively, guarding the call to `PosixRename` with a check for the extension and using `s.cli.Rename(tempFile, fullPath)` (without removing the file beforehand) if the extension is not supported would be the other option with the smallest code impact. I am trying to stay away from calling remove on the file, because it doesn't seem like a clean solution when we know that that call will fail, because the file does not exist, though it would be another option and should work just fine in practice.

Generally, there seems to be a slight difference between what the tests are asserting (i.e. the ability to overwrite blobs) and what Kopia seems to be requiring (i.e. just creating new blobs), but I don't think I am qualified enough to have an opinion on that.

I would like to hear your thoughts on this.

**Comment by jkowalski** [comment]
Yes, I think the test is overly aggressive and honestly probably needs to be rewritten. I think removing overwrite logic makes sense for now.

The logic for non-posix servers could be:

```
while already_exists(Rename(old,new)) {
  Remove(new)
}
```

for Posix servers I would keep it as-is.

BTW. This SFTP could use some additional improvements and if you have time to hack on it, I'd really appreciate that. The biggest issue is the ability to reconnect to the server - at this point when the server is disconnected, the entire session is effectively busted with no way of recovering.
